File: scrape.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/PuerkitoBio/goquery"
)

func getCount(client *http.Client, language *Lang, wg *sync.WaitGroup) {
	defer wg.Done()

	req, err := http.NewRequest("GET", language.Url, nil)
	throw(err)
	req.Header.Add("User-Agent", `Mac:com.mikekreuzer.ripley:0.2.0 (by /u/mikekreuzer)`)

	resp, err := client.Do(req)
	throw(err)
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		doc, err := goquery.NewDocumentFromReader(resp.Body)
		throw(err)

		countAsString := doc.Find("span.subscribers").First().Find("span.number").First().Text()
		if countAsString != "" {
			count, err := strconv.Atoi(strings.Replace(countAsString, ",", "", -1))
			throw(err)
			language.Subscribers = count
			language.SubsString = countAsString
		}
	} else {
		fmt.Println("%s - %s", http.StatusText(resp.StatusCode), language.Url)
	}
}

func writeBytes(fileName string, b []byte) {
	f, err := os.Create(fileName)
	throw(err)
	defer f.Close()
	_, err = f.Write(b)
	throw(err)
}

func outputTotal(languages []*Lang, tableSize int) {
	sort.Sort(sort.Reverse(BySubs(Languages)))

	total := 0
	for i, language := range languages {
		language.Index = i + 1
		total += language.Subscribers
	}

	for _, language := range languages {
		perc := float64(language.Subscribers) / float64(total) * 100.0
		language.Percentage = strconv.FormatFloat(perc, 'f', 1, 64)
	}

	// json
	theDate := time.Now()
	data := &Output{theDate.Format("January 2006"), theDate.Format(time.RFC3339), languages}
	b, err := json.Marshal(data)
	throw(err)
	var out bytes.Buffer
	json.Indent(&out, b, "", " \t")

	// file name
	fileNameParts := []string{"new-", theDate.Format("01-2006"), ".json"}
	fileName := path.Join("data", strings.Join(fileNameParts, ""))

	// write file
	writeBytes(fileName, out.Bytes())
}

File: scrape.rb
require 'concurrent'
require 'date'
require 'json'
require 'mechanize'
require_relative './languages.rb'

def insert_percentages_and_index(array_of_hashes)
  total = array_of_hashes.map { |r| r[:subscribers] }.reduce(0, :+)
  return array_of_hashes if total.zero?
  results = array_of_hashes.map.with_index do |lang, index|
    lang[:percentage] = (lang[:subscribers].to_f / total * 100).round(2)
    lang[:index] = index + 1
    lang
  end
  results
end

# scrape the language pages
workers = []
LANGUAGES.each do |language, url|
  workers << Concurrent.dataflow do
    agent = Mechanize.new
    agent.request_headers = {
      'User-Agent' => 'Mac:com.mikekreuzer.ripley:0.8.0 (by /u/mikekreuzer)'
    }
    page = agent.get(url)
    count = page.at('.subscribers .number').text.strip
    { name: language, subsstring: count, subscribers: count.delete(',').to_i }
  end
end

# tally the results
tally = Concurrent.dataflow(*workers) do
  results = []
  workers.each { |work| results << work.value }
  sorted_results = results.sort_by { |hsh| hsh[:subscribers] }.reverse!
  results = insert_percentages_and_index(sorted_results)
  results
end

date = DateTime.now

result = { 'title' => date.strftime('%B %Y'),
           'dateScraped' => date,
           'data' => tally.value }

File.write date.strftime('%m-%Y.json'), JSON.pretty_generate(result)

# add in comparison to previous...
# add in moving mmd file to zine, & calling zine
